src/libcore/num/mod.rs

```diff
@@ -1471,3 +1472,73 @@ impl fmt::Display for ParseIntError {
 }
 
 pub use num::dec2flt::ParseFloatError;
+
+
+// Conversion traits for primitive integer types
+// Conversions T -> T are covered by a blanket impl and therefore excluded
+macro_rules! impl_from {
+    ($Small: ty, $Large: ty) => {
+        impl From<$Small> for $Large {
+            #[stable(feature = "lossless_int_conv", since = "1.5.0")]
+            #[inline]
+            fn from(small: $Small) -> $Large {
+                small as $Large
+            }
+        }
+    }
+}
+
+// Unsigned -> Unsigned
+impl_from! { u8, u16 }
+impl_from! { u8, u32 }
+impl_from! { u8, u64 }
+impl_from! { u8, usize }
+impl_from! { u16, u32 }
+impl_from! { u16, u64 }
+impl_from! { u16, usize }
+impl_from! { u32, u64 }
+#[cfg(any(target_pointer_width = "64", target_pointer_width = "32"))]
+impl_from! { u32, usize }
+#[cfg(target_pointer_width = "64")]
```

Comment:
>Well, currently all this code that would fail to compile with these changes is silently truncating the integers. I prefer a compile-time error to that.

I share  this sentiment.
In practice, there's not much difference between current situation and platform-dependent `into` when you target 32 and 64. If you keep usual level of "awareness of portability problems" you can avoid `into` in the same places where you would avoid potentially problematic `as`. As I said, I methodically replaced `as` in rustc+libstd with lossless `into` and checked `cast` (keeping 32<->64 portability in mind) on 64-bit Windows and then everything "just worked" on 32-bit Linux with exception of few places, that were "interesting" and actually required human attention. So, there *is* some experience dispelling fears about potential portability hazards. If someone is especially inaccurate and writes totally non-portable 32bit-only (or 64bit-only) code, then an easy porting solution also exists - write a regexp and just replace all `into`s with `cast`s.
(16 bit is a completely separate story and I think they will appreciate a compile time error when something shrinks from 4294967296 to 65536)